mmlearn/tasks/linear_probing.py

```diff
@@ -0,0 +1,454 @@
+from typing import Any, Dict, List, Literal, Optional, Tuple, Union
+
+import hydra
+import lightning as L  # noqa: N812
+import torch
+from hydra_zen import store
+from lightning.pytorch.utilities.types import OptimizerLRScheduler
+from lightning_utilities.core.rank_zero import rank_zero_warn
+from torch import nn
+from torchmetrics import AUROC, Accuracy, F1Score, MetricCollection, Precision, Recall
+
+from mmlearn.datasets.core import Modalities
+from mmlearn.modules.layers import MLP
+
+
+def extract_vision_encoder(
+    encoder: Any,
+    model_checkpoint_path: Optional[str],
+    keys_to_remove: Optional[List[str]] = None,
+    keys_to_rename: Optional[Dict[str, str]] = None,  # Default for renaming
+    keys_to_ignore: Optional[List[str]] = None,
+) -> nn.Module:
+    """
+    Extract the vision encoder from a PyTorch Lightning model.
+
+    Args:
+        encoder (Any): The encoder configuration or model to be instantiated.
+        model_checkpoint_path (Optional[str]): Path to the checkpoint file containing
+        the encoder's state_dict.
+        keys_to_remove (Optional[list]): List of keys to be removed from the state_dict.
+        keys_to_rename (Optional[dict]): Dictionary of prefixes or key replacements
+        mapping
+            old prefixes to new replacements (default removes 'encoders.rgb.').
+        keys_to_ignore (Optional[list]): List of keys to ignore when loading the
+        state_dict.
+
+    Returns
+    -------
+        nn.Module: The vision encoder module extracted and initialized.
+    """
+    model: L.LightningModule = hydra.utils.instantiate(encoder)
+    if model_checkpoint_path is None:
+        rank_zero_warn(
+            "No model_checkpoint_path path was provided for linear evaluation."
+        )
+    else:
+        checkpoint = torch.load(model_checkpoint_path)
+        if "state_dict" not in checkpoint:
+            raise KeyError("'state_dict' not found in checkpoint")
+
+        state_dict = checkpoint["state_dict"]
+
+        # Remove unwanted keys
+        if keys_to_remove:
+            state_dict = {
+                k: v for k, v in state_dict.items() if k not in keys_to_remove
+            }
+
+        # Ignore specific keys
+        if keys_to_ignore:
+            state_dict = {
+                k: v for k, v in state_dict.items() if k not in keys_to_ignore
+            }
+
+        # Rename keys based on input mappings
+        if keys_to_rename:
+            state_dict = {
+                k.replace(old_prefix, new_prefix): v
+                for k, v in state_dict.items()
+                for old_prefix, new_prefix in keys_to_rename.items()
+                if k.startswith(old_prefix)
+            }
+
+        try:
+            if state_dict:
+                model["rgb"].load_state_dict(state_dict, strict=True)
+                print("Encoder state dict loaded successfully")
+        except Exception as e:
+            print(f"Error loading state dict: {e}")
+    return model["rgb"]
```

Comment:
Please generalize this as well. That is, allow the user to specify which key, if any, to extract.